Add output to New task with Color

yez opened the pull request:
Introduce a Color.colorize method with a few default colors that allow
the output to be in a few bash colors. Happy to add more colors as well!

Closes #33

There was already some output so not sure what you'd like the output to say, here is an example of the current output. 

<img width="1147" alt="Screen Shot 2020-10-10 at 8 21 05 PM" src="https://user-images.githubusercontent.com/638533/95667615-75f40d00-0b36-11eb-8de6-8526321357b9.png">

Commit:
Add output to New task with Color

Introduce a Color.colorize method with a few default colors that allow
the output to be in a few bash colors.

Closes #33

yez commented:
Looks like something failed but unfortunately clicking on the `ci/ build` links don't show much.

juanmanuelramallo commented:
Yeah the UI is quite hard to follow.

But here's what failed https://github.com/rootstrap/rsgem/pull/55/checks?check_run_id=1236805720#step:11:1

```
Run bundle exec rake code_analysis
bundle exec rubocop lib spec
Inspecting 36 files
.....C.C............................

Offenses:

lib/rsgem/cli/commands/new.rb:26:9: C: Metrics/AbcSize: Assignment Branch Condition size for call is too high. [<0, 16, 1> 16.03/15]
        def call(**options) ...
        ^^^^^^^^^^^^^^^^^^^
lib/rsgem/cli/commands/new.rb:29:88: C: Style/IfUnlessModifier: Modifier form of unless makes the line too long.
          puts Colors.colorize("\tpassed in bundler option = #{options[:ci]}", :green) unless options[:bundler].nil?
                                                                                       ^^^^^^
lib/rsgem/cli/commands/new.rb:29:101: C: Layout/LineLength: Line is too long. [116/100]
          puts Colors.colorize("\tpassed in bundler option = #{options[:ci]}", :green) unless options[:bundler].nil?
                                                                                                    ^^^^^^^^^^^^^^^^
lib/rsgem/colors.rb:16:20: C: Style/TrailingCommaInHashLiteral: Avoid comma after the last item of a hash.
          white: 97,
```

Rubocop being rubocop see_no_evil but I guess these are valid concerns to keep the same code style in the gem.

yez commented:
Thanks for pointing this out @juanmanuelramallo, I'll be sure to run rubocop before committing again in this and my other PR

juanmanuelramallo commented on lib/rsgem/cli/commands/new.rb:
```diff
@@ -24,6 +24,9 @@ class New < Dry::CLI::Command
         ]
 
         def call(**options)
+          puts Colors.colorize("Creating a new gem #{options[:gem_name]}", :green)
+          puts Colors.colorize("\tpassed in ci option = #{options[:ci]}", :green)
+          puts Colors.colorize("\tpassed in bundler option = #{options[:ci]}", :green) unless options[:bundler].nil?
```
I'd rather leave the output information outside the CLI definition. The CLI related objects should only care about defining the valid options/arguments and mapping them into a method call of the gem, like we do in here for the new command. 

What do you think if instead we only update (and maybe colorize) these type of messages:
- https://github.com/rootstrap/rsgem/blob/master/lib/rsgem/tasks/create_gem.rb#L10
- https://github.com/rootstrap/rsgem/blob/master/lib/rsgem/tasks/clean_gemfile.rb#L21
- https://github.com/rootstrap/rsgem/blob/master/lib/rsgem/tasks/ignore_gemfile_lock.rb#L14

maybe any `puts` call in task objects in general (?)

yez replied:
I thought about this after I submitted the PR, I agree that tasks should do the I/O and not this CLI object. I'll refactor this a bit.

juanmanuelramallo commented on lib/rsgem/colors.rb:
```diff
@@ -0,0 +1,21 @@
+      end
+
+      def color_code(color)
+        {
+          red: 31,
+          green: 32,
+          yellow: 33,
+          blue: 34,
+          white: 97,
+        }[color] || 39
```
Maybe we can make this hash a constant in this class to avoid allocating memory for it each time we colorize something (I know it shouldn't matter the memory consumption, so maybe from a code style perspective a think a constant would be more suitable in here)

```ruby
module RSGem
  class Colors
    COLORS = {
      red: 31,
      # ...
    }.freeze

    ...

    def color_code(color)
      COLORS[color] || 39
    end
```

Something like this ^^^ maybe (?)

yez replied:
Yeah I like that better, I'll change it to a constant. Any other colors you think we should add while I'm here?

yez committed:
Fix Rubocop and Output in Tasks

Update colors for existing task output, refactor Color class to use
a constant of colors with a default.

juanmanuelramallo reviewed (approved):
Amazing! Let's add an entry in the changelog for this is as well please! Below master will be just fine.

yez committed:
Update Changelog

yez commented:
@juanmanuelramallo Rebased and Updated the CHANGELOG!

juanmanuelramallo commented:
Amazing thank you!
